src/common/datasource/src/file_format/orc.rs

```diff
@@ -123,3 +124,40 @@ impl FileFormat for OrcFormat {
         Ok(schema)
     }
 }
+
+#[derive(Debug, Clone)]
+pub struct OrcOpener {
+    object_store: Arc<ObjectStore>,
+    schema: SchemaRef,
+}
+
+impl OrcOpener {
+    pub fn new(object_store: ObjectStore, schema: SchemaRef) -> Self {
```

Comment:
Add some comments for the new function, and the `schema` seems better to be called the `projected schema`?

Reply:
Considers using `projection: Vec<usize>` instead of the `schema: SchemaRef`. It can help to reduce the complexity of implementing the stream projector.

Reply:
@WenyXu Are you suggesting to create OrcOpener with projection: Vec<usize> instead of schema: SchemaRef? If so, it seems inconsistent with the parameters required for creating OrcArrowStreamReaderAdapter.

Reply:
> @WenyXu Are you suggesting to create OrcOpener with projection: Vec instead of schema: SchemaRef? If so, it seems inconsistent with the parameters required for creating OrcArrowStreamReaderAdapter.

Yep, And also adapt the `OrcArrowStreamReaderAdapter ` for your changes, e.g., Add a `projection parameter for `OrcArrowStreamReaderAdapter`.

Reply:
@WenyXu 
I'm not sure if I misunderstood, but why does the creation of orc_opener only require the projection and object_store parameters, and not the schema parameter? In my understanding, even if the projection is provided, the schema is still needed to determine the projected_schema.

This is my understanding:

```rust
pub struct OrcOpener {
    object_store: Arc<ObjectStore>,
    file_schema: SchemaRef,
    projection: Option<Vec<usize>>
}

pub struct OrcArrowStreamReaderAdapter<T: AsyncRead + AsyncSeek + Unpin + Send + 'static> {
    file_schema: SchemaRef,
    projection: Vec<usize>,
    stream: ArrowStreamReader<T>,
}
```
In my understanding, even if the projection parameter is provided, the schema parameter is still needed to determine the projected_schema.

In the OrcArrowStreamReaderAdapter, the projection parameter is used to optimize the retrieval of the corresponding columns, and the file_schema parameter is necessary to generate the projected_schema.

Is my understanding correct?

Reply:
Yes, you are right. However, `orc` is a self-description format.  You can retrieve the schema by calling the `schema` method.

```rust
            let stream_reader = new_orc_stream_reader(reader)
                .await
                .map_err(|e| DataFusionError::External(Box::new(e)))?;

            let schema = stream_reader.schema();
```

Reply:
> Yes, you are right. However, `orc` is a self-description format. You can retrieve the file schema by calling the `schema` method.
> 
> ```rust
>             let stream_reader = new_orc_stream_reader(reader)
>                 .await
>                 .map_err(|e| DataFusionError::External(Box::new(e)))?;
> 
>             let schema = stream_reader.schema();
> ```

Yes, but if we do it this way, it seems like it won't be able to be converted into the required type.  
[fix: cast orc data against output schema #1922 ](https://github.com/GreptimeTeam/greptimedb/pull/1922).So it seems like we still need an output_schema, and it appears that the projection is not necessary.

Reply:
> > Yes, you are right. However, `orc` is a self-description format. You can retrieve the file schema by calling the `schema` method.
> > ```rust
> >             let stream_reader = new_orc_stream_reader(reader)
> >                 .await
> >                 .map_err(|e| DataFusionError::External(Box::new(e)))?;
> > 
> >             let schema = stream_reader.schema();
> > ```
> 
> Yes, but if we do it this way, it seems like it won't be able to be converted into the required type. [fix: cast orc data against output schema #1922 ](https://github.com/GreptimeTeam/greptimedb/pull/1922).So it seems like we still need an output_schema, and it appears that the projection is not necessary.



Let me explain schemas. 
- The `output_schema`  is what the caller expected. (`OrcArrowStreamReaderAdapter` converts data against the `output_schema`)
- The `file_schema` describes the file data types.

We need `output_schema`, which converts data (`file_schema`) against the `output_schema`.

Therefore, The code should look like the following:

```rust
pub struct OrcOpener { // OrcOpener able to retrieve the file schema from the reader. Therefore, we don't need the file_schema field.
    object_store: Arc<ObjectStore>,
    projection: Option<Vec<usize>>
}

pub struct OrcArrowStreamReaderAdapter<T: AsyncRead + AsyncSeek + Unpin + Send + 'static> {
    output_schema: SchemaRef,
    projection:  Option<Vec<usize>>,
    stream: ArrowStreamReader<T>,
}
```

Reply:
The file table engine only allows users to define the custom table schema(and store it in the table metadata) for non-self-description formats (e.g., JSON, CSV); the underlying reader will cast the string data into the target type.  

For self-description formats (parquet, orc), the table metadata 's `file_schema` must be against the file's schema.

Therefore, In this scenario, the `file schema` in the file table engine actually equals the `output schema`.

Reply:
> The file table engine only allows users to define the custom table schema(and store it in the table metadata) for non-self-description formats (e.g., JSON, CSV); the underlying reader will cast the string data into the target type.
> 
> For self-description formats (parquet, orc), the table metadata 's `file_schema` must be against the file's schema.
> 
> Therefore, In this scenario, the `file schema` in the file table engine actually equals the `output schema`.

Understood, thank you for your explanation.